File: product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] ans = new int[nums.length];
        if (nums.length == 0) {
            return ans;
        }
        
        ans[0] = 1;
        for (int i = 1; i < nums.length; i++) {
            ans[i] = ans[i - 1] * nums[i - 1];
        }
        
        int rightProduct = 1;
        for (int i = nums.length - 2; i >= 0; i--) {
            rightProduct *= nums[i + 1];
            ans[i] *= rightProduct;
        }
        
        return ans;
    }
}

File: product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = [1] * len(nums)
        
        left_product = nums[0]
        for i in range(1, len(nums)):
            res[i] *= left_product
            left_product *= nums[i]
            
        right_product = nums[-1]
        for i in range(len(nums) - 2, -1, -1):
            res[i] *= right_product
            right_product *= nums[i]
        
        return res
        